### DISCUSSION TITLE
About get remote ip&port

### DISCUSSION THREAD

**Discussion by cloudandzack007** [OpenDDS/OpenDDS#3726] [Q&A] [answered]
When the function [ virtual void on_subscription_matched(
        DDS::DataReader_ptr reader,
        const DDS::SubscriptionMatchedStatus& status); ] is called, how to get the ip&port by last_publication_handle, i can only get SubscriptionMatchedStatus and qospolicy. 
I read the source code and find the remote_address of TcpConnection, but no interface to get them.

**Comment by jrw972** [comment, marked as answer]
I would first consider what you are trying to do.

If you are looking for some way of unambiguously identifying the remote publication, then you can take the handle and look up the publication in the Publication Builtin Topic Datareader.  The key on the returned sample will be a 16-byte guid that identified the remote datawriter.

Since OpenDDS guids are dynamic, you may want to use identifiers from your application.  If this is the case, then use the `user_data` qos of the datawriter and look it up using the approach described above.

If you are really interested in the ip:port then you have some options.

The ParticipantLocationTopic is an experimental feature that reports the address:port for SPDP discovery.
The ConnectionRecordTopic is an incomplete generalization of the ParticipantLocationTopic that will give you what you want.
(You will need to look up the GUID first.)

Finally, you could add the necessary plumbing to the create the API you desire.
You have two choices here as well.
First, you could add to the Discovery interface `dds/DCPS/Discovery.h` to return the transport locator sequence for a remote publication.  You could then decode the transport locator sequence which should be treated as a list of candidates.  Second, you could add to `dds/DCPS/DataReaderImpl.h` to return a transport locator for a remote publication.  This would be the actual address being used.  DataReaderImpl is a transport client which has a relationship with a transport impl which has a relationship with a datalink which knows the address.

**Reply by cloudandzack007** [reply]
thank you very much for your answer
